text/0000-projections-lifetimes-and-wf.md

```diff
@@ -0,0 +1,980 @@
+implied bounds and contravariance ([#25860]). To better separate the
+issues, this will be addressed in a follow-up RFC that should appear
+shortly.
+
+#### Missing WF checks
+
+Unfortunately, the compiler currently fails to enforce WF in several
+important cases. For example, the
+[following program](http://is.gd/6JXjyg) is accepted:
+
+```rust
+struct MyType<T:Copy> { t: T }
+
+trait ExampleTrait {
+    type Output;
+}
+
+struct ExampleType;
+
+impl ExampleTrait for ExampleType {
+    type Output = MyType<Box<i32>>;
+    //            ~~~~~~~~~~~~~~~~
+    //                   |
+    //   Note that `Box<i32>` is not `Copy`!
+}
+```
+
+However, if we simply naively add the requirement that associated
+types must be well-formed, this results in a large annotation burden
+(see e.g. [PR 25701](https://github.com/rust-lang/rust/pull/25701/)).
+For example, in practice, many iterator implementation break due to
+region relationships:
+
+```rust
+impl<'a, T> IntoIterator for &'a LinkedList<T> { 
+   type Item = &'a T;
+   ...
+}
+```
+
+The problem here is that for `&'a T` to be well-formed, `T: 'a` must
+hold, but that is not specified in the where clauses. This RFC
+proposes using implied bounds to address this concern -- specifically,
+every impl is permitted to assume that all types which appear in the
+impl header (trait reference) are well-formed, and in turn each "user"
+of an impl will validate this requirement whenever they project out of
+a trait reference (e.g., to do a method call, or normalize an
+associated type).
+
+# Detailed design
+
+This section dives into detail on the proposed type rules.
+
+### A little type grammar
+
+We extend the type grammar from [RFC 192] with projections and slice
+types:
+
+    T = scalar (i32, u32, ...)        // Boring stuff
+      | X                             // Type variable
+      | Id<P0..Pn>                    // Nominal type (struct, enum)
+      | &r T                          // Reference (mut doesn't matter here)
+      | O0..On+r                      // Object type
+      | [T]                           // Slice type
+      | for<r..> fn(T1..Tn) -> T0     // Function pointer
+      | <P0 as Trait<P1..Pn>>::Id     // Projection
+    P = r                             // Region name
+      | T                             // Type
+    O = for<r..> TraitId<P1..Pn>      // Object type fragment
+    r = 'x                            // Region name
+    
+We'll use this to describe the rules in detail.
+    
+### Syntactic definition of the outlives relation
+
+The outlives relation is defined in purely syntactic terms as follows.
+These are inference rules written in a primitive ASCII notation. :) As
+part of defining the outlives relation, we need to track the set of
+lifetimes that are bound within the type we are looking at.  Let's
+call that set `R=<r0..rn>`. Initially, this set `R` is empty, but it
+will grow as we traverse through types like fns or objects, which can
```

Comment:
nit: could/should this say  something like "fn types or object-type fragments, which are lifetime-parametrized via `for <..> T` ?  At first I was skimming down to see if "objects" did indeed add to the `R`-set, but the only rule that has "Object" in its name ("OutlivesObject") does not add to the `R`-set.

Then I realized that by "objects" you meant the "Object type fragments" as listed above.  Which makes sense since that is indeed a binding form for region names.

I don't know, I basically am saying that referencing the `for <..> T` form here might help.